cql3: forbid creating MVs with gc_grace_seconds=0

piodul opened the pull request:
Like in the case of Scylla, in Cassandra's implementation of MVs, base replicas are responsible for generating view updates and delivering them to the paired view replicas. The view update is stored in a batch log for better durability guarantees, and asynchronously replayed in the background. This is slightly different to Scylla's behavior, which just attempts to send the view update as if it were doing a write, and stores a hint if the write fails.

Entries in the batchlog are expired after gc_grace_seconds, so it doesn't make sense to set it to 0 for materialized views because the entries are expired immediately but never delivered. In case of Scylla, gc_grace_seconds=0 also is not a good idea because hints expire in a very similar way - this effectively disables view hints.

In CASSANDRA-9917, some checks were implemented which, specifically, forbid altering a view with gc_grace_seconds=0. However, it is possible to create a new materialized view with gc_grace_seconds=0 - which doesn't make sense for Cassandra's implementation, and is problematic for our implementation.

Let's implement the missing check for CREATE.

Fixes: scylladb/scylladb#20164

Not a regression and not a high impact bug at all, no need to backport in my opinion.

piodul commented on cql3/statements/create_view_statement.cc:
```diff
@@ -164,6 +164,15 @@ std::pair<view_ptr, cql3::cql_warnings_vec> create_view_statement::prepare_view(
+                "of 0 for base table '{}' , since this value is "
+                "used to TTL undelivered updates. Setting gc_grace_seconds "
+                "too low might cause undelivered updates to expire "
+                "before being replayed.", column_family(), _base_name.get_column_family()));
```
> Note about the commit message: you say "The view update is stored in a batch log for better durability guarantees". This doesn't actually happen.

It doesn't happen in Scylla, but it does happen in Cassandra. I was basing my comment on what I understood from http://www.doanduyhai.com/blog/?p=1930. Perhaps I should have been more clear that I'm speaking about Cassandra there, but I thought the first paragraph makes it clear.

> 
> Maybe you're talking about _hints_, stored only when a remote node is down?
> 
> But in that case, what actually gets GC'ed just the tombstones? If we expire tombstones that we store in hints, it sounds like a bug in hints, not something we should hide in the MV code :-(

Unsent hints are garbage-collected after gc_grace_seconds, regardless whether they are tombstones or not - that's just a part of their design. I think this behavior makes sense for the general use case at least - without the GC behavior the following case can happen:

- Hint to partition P is stored on node A,
- Partition P is deleted on node B by a tombstone with higher timestamp than the hint,
- gc_grace_seconds pass, the tombstone on node B gets garbage collected,
- Hint gets replayed by A and sent to B - this leads to data resurrection.

Cassandra batchlog behaves similarly.

nyh replied:
It did confuse me - I didn't understand that much of the text in the commit message is actually about Cassandra, not Scylla. I'm not sure, by the way, why we need to explain here how Cassandra works, but I guess if you already researched it it can't hurt to include it - but perhaps you can just reorganize it to make it clearer where you're talking about Scylla - which has HINTS - and where you're talking about Cassandra that has other stuff.

By the way, I'm at a loss why we don't just fix the hints bug instead of working around it. Hints are like repair - you need to apply whatever comes through it - both live data and tombstones. You're not supposed to GC tombstones you get in hints before even applying them...

piodul replied:
> You're not supposed to GC tombstones you get in hints before even applying them...

It's not that we expire tombstones in hints, we completely drop hints that are older than gc_grace_seconds, no matter whether they hold tombstones live data.

As for whether it's a good idea - I'm not entirely sure. Hints are not guaranteed to be reliable, so dropping them after gc_grace_seconds might be a better idea than dropping them because it avoids data resurrection.

nyh replied:
Well, maybe it's a bug to "completely drop hints that are older than gc_grace_seconds"? How the @#!? is gc_grace_seconds related to dropping old hints? If I set gc_grace_seconds to 1 second (because you don't allow me to set it to 0), how will I not have problems now that all hints supposedly get discarded after 1 second?
I understand that we might have various rules for dropping hints, but I don't see why they should be paramterized on gc_grace_seconds, which has nothing to do with hints.
Is this just a bug? Shouldn't we fix it instead of this patch trying to paper over this bug in a really strange way?

bhalevy replied:
The gc grace period model assumed that repair will run in the grace period, and repair is supposed to flush hints, right?
And the hints are not needed after repair.
The bottom line is that repair should be the mechanism to "trim" hints, not a time-based mechanism.
That's why tombstone_gc_mode=repair is so important.

That said, we also have the odd cases of tombstone_gc_mode=immediate or timeout mode and gc_grace_seconds=0, where we can't expect repair to run before garbage collection.  That may be suitable when using rf=1.

So what about the hints in this case?
Is this a problem only for views?
Why isn't it a problem for regular tables as well?
Maybe we should disable hints (and view hint) for tables using immediate tombstone_gc alltogether?

nyh commented on test/cql-pytest/test_materialized_view.py:
```diff
@@ -288,6 +288,18 @@ def test_is_not_null_forbidden_in_filter(cql, test_keyspace, cassandra_bug):
             cql.execute(f"DROP MATERIALIZED VIEW IF EXISTS {test_keyspace}.{mv}")
 
+# Test that it is not allowed to create a materialized view with gc_grace_seconds=0.
+# This check is missing from Cassandra, so the test is marked as Scylla-only.
```
Thanks. Please explain in the comment before the test - not just in the cover letter - why you consider it to be a bug in Cassandra, and use the fixture "cassandra_bug" - not "scylla_only". (and please also mention that this test reproduces the issue number)

If you feel like it, it will also be nice to open an issue on Cassandra explaining how what they implemented in CASSANDRA-9917 is inconsistent (checks only ALTER but not CREATE). Then you'll get the official stamp of approval that it's a Cassandra bug ;-) But it's not required, only do it if you want to be a good Cassandra citizen.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11633/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11633/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_materialized_view
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11633/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/572/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11642/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3625/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11633/)
#### Failed Tests (18/49229):
- [test_tombstone_gc_with_materialized_view_update_in_memtable.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.debug.1)
- [test_tombstone_gc_with_materialized_view_update_in_memtable.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.debug.2)
- [test_tombstone_gc_with_materialized_view_update_in_memtable.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.debug.3)
- [test_tombstone_gc_with_materialized_view_update_in_memtable.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.dev.1)
- [test_tombstone_gc_with_materialized_view_update_in_memtable.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.dev.2)
- [test_tombstone_gc_with_materialized_view_update_in_memtable.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.dev.3)
- [test_tombstone_gc_with_materialized_view_update_in_memtable.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_release_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.release.1)
- [test_tombstone_gc_with_materialized_view_update_in_memtable.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_release_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.release.2)
- [test_tombstone_gc_with_materialized_view_update_in_memtable.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_release_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.release.3)
- [cql-pytest.test_compaction_tombstone_gc.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_test_compaction_tombstone_gc_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.test_compaction_tombstone_gc.debug.1)
#### Build Details:
- Duration: 3 hr 27 min
- Builder: spider6.cloudius-systems.com

piodul commented:
> #### Failed Tests (18/49229):
> * [test_tombstone_gc_with_materialized_view_update_in_memtable.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.debug.1)
> * [test_tombstone_gc_with_materialized_view_update_in_memtable.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_debug_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.debug.2)
> * [test_tombstone_gc_with_materialized_view_update_in_memtable.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_debug_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.debug.3)
> * [test_tombstone_gc_with_materialized_view_update_in_memtable.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_dev_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.dev.1)
> * [test_tombstone_gc_with_materialized_view_update_in_memtable.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_dev_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.dev.2)
> * [test_tombstone_gc_with_materialized_view_update_in_memtable.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_dev_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.dev.3)
> * [test_tombstone_gc_with_materialized_view_update_in_memtable.release.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_release_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.release.1)
> * [test_tombstone_gc_with_materialized_view_update_in_memtable.release.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_release_2) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.release.2)
> * [test_tombstone_gc_with_materialized_view_update_in_memtable.release.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/test_compaction_tombstone_gc/Tests___Unit_Tests___test_tombstone_gc_with_materialized_view_update_in_memtable_release_3) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tombstone_gc_with_materialized_view_update_in_memtable.release.3)
> * [cql-pytest.test_compaction_tombstone_gc.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/11633/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___cql_pytest_test_compaction_tombstone_gc_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+cql-pytest.test_compaction_tombstone_gc.debug.1)

It looks like a scylla-only test was introduced recently which uses a view with gc gs = 0... It doesn't seem required there, so I'll fix it.

piodul commented:
v2:
- Reworded the commit message
- Added an explanatory comment to test_zero_gc_grace_seconds_for_view_is_forbidden, marked it as cassandra_bug and not scylla_only
- Fixed the tests in test_compaction_tombstone_gc.py which used to create a view with gc_grace_seconds=0

piodul commented on test/cql-pytest/test_compaction_tombstone_gc.py:
```diff
@@ -9,7 +9,7 @@
 # the commitlog minimum gc time, in seconds,
 # to be greater than the tombstone deletion time
-def sleep_till_whole_second(seconds=1):
+def sleep_till_whole_second(seconds=2):
```
@bhalevy could you take a look at the first commit (`cql-pytest: don't use gc_grace_seconds=0 for MVs`)?

nyh replied:
This patch is saddening for several reasons.
First there is the sleeps.
Second, it shows how silly it is to forbid a gc_grace_seconds=0 but allow gc_grace_seconds=1.
Thirdly, when this test was written by @bhalevy he told me, which really surprised me, that gc_grace_seconds=0 actually means wait until the next second, not zero seconds. So why do we have a problem with zero seconds.
Finally, I still don't understand why exactly we can't use gc_grace_seconds=0. Because of a bug with hints? Couldn't we fix this bug?

nyh replied:
Ok, I finally understand what's going on with gc_grace_seconds (my "finally" above), see my other comments. I still feel like we need to fix the underlying bug (or a really really awkward feature) instead of pretending that it no longer exists if you're not allowed to set gc_grace_seconds=0 but can still set gc_grace_seconds=1.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11834/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11834/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: cql-pytest/test_compaction_tombstone_gc
:small_blue_diamond: cql-pytest/test_materialized_view
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11834/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/745/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/3800/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/11815/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/11834/)
#### Build Details:
- Duration: 8 hr 13 min
- Builder: i-0e4121bfcb9a60a41 (m5ad.8xlarge)

nyh commented on test/cql-pytest/test_materialized_view.py:
```diff
@@ -288,6 +288,30 @@ def test_is_not_null_forbidden_in_filter(cql, test_keyspace, cassandra_bug):
+# to replay failed view updates. In both cases, batchlog entries / hints expire
+# after gc_grace_seconds of the view. Therefore, it doesn't make sense to set
+# gc_grace_seconds to a small value as this causes entries / hints to expire
+# immediately.
```
I found this text hard to believe, so I looked, and, you're right, Cassandra does exactly this crazy thing. I even found a document on it - https://thelastpickle.com/blog/2018/03/21/hinted-handoff-gc-grace-demystified.html - but I still don't understand WHY they do it.

But now I don't understand - this crazy behavior breaks not just gc_grace_seconds but any "small" gc_grace_seconds - even one hour is too little if there is a chance that hints need to be replayed after one hour.

Also, when hints are enabled on a base table - not a view, we should have the same problem! Why is this check only for views?

It sounds to me like we should either:
1. Rethink this weird gc_grace_seconds-used-to-expire-hints thing.
or
2. Add a "guardrails"-like configuration parameter of "minimum_gc_grace_seconds", with lower numbers causing an error. We could set the default minimum to, say, one day, but allow us to use 0 as the guardrail setting in tests to make these tests fast.

bhalevy replied:
As I said [above](https://github.com/scylladb/scylladb/pull/20618#discussion_r1774355613), the tombstone gc grace period, is based on having repair run in the grace period to ensure data and tombstones are synchronized, making tombstone garbage collection safe.
And repair should flush hints (and batchlog).
Sadly, repair does not run when rf=1, making immediate tombstone garbage collection somewhat sensible in this configuration.  But this might expose a hole in the design if hints or batchlog may replay deleted data after tombstones were gc'ed.

bhalevy commented:
> Fixes: https://github.com/scylladb/scylladb/issues/20164

This PR doesn't really fix the issue.
It should use `Closes` rather than `Fixes`
(but I'm afarid it's not the right direction anyhow...)

nyh commented:
Hi @piodul I'm trying to document the gc_grace_seconds story in my new MV document (https://docs.google.com/document/d/1yk7MVkXM3LEDV_sb9FMHCQmMJYYwGN6fwsz3YzbOEc0/edit#) and I'm not sure I understood the whole story. If I understand correctly, the reason why we don't want gc_grace_seconds=0 for MV is that in the current implementation, *hints* get expired if they are older than gc_grace_seconds. But my question is: **view hints** are special, they are not ordinary hints for the base table. When we wrtite them, which expiration time do we use for them - the base's gc_grace_seconds or the view's? Do you know where the code that does this lives?

I belive that in Cassandra, their version of these hints (batch log) use's the base's gc_grace_seconds for expiration, which is why Cassandra didn't want the gc_grace_seconds of the **base** to be 0 (or actually, anything too small will be a problem...) - so this is what they did in https://issues.apache.org/jira/browse/CASSANDRA-9917. Do we know for sure that in Scylla's implementation, the view hints take their gc_grace_seconds from the view table, not the base?

nyh commented:
> > Fixes: #20164
> 
> This PR doesn't really fix the issue.

I read https://issues.apache.org/jira/browse/CASSANDRA-9917 again, I think I know in what sense this patch "Fixes" the inconsistency described issue (if you think the issue doesn't describe the inconcistency well you can close it and open a better one...):

In the original https://issues.apache.org/jira/browse/CASSANDRA-9917 it was noted that any small gc_grace_seconds will be bad, because it can cause view hints (in their case - view batchlog) to expire before they are replayed. So the original patch they proposed a constant MIN_GC_GRACE_SECONDS = 3 hours, and forbid gc_grace_seconds to be lower than that if a view is involved. However, a debate started what this constant should be set to and what does it even guarantee (someone gave an example where two nodes each are down for just 1.5 hours but causing the hints not to be replayed for 3 hours). Eventually there was a decision by one of the developers that nobody ever sets gc_grace_seconds to 7 hours or 1.5 days - users almost always use either the default (10 days) or 0. So we just need to forbid the 0.

piodul commented:
> Hi @piodul I'm trying to document the gc_grace_seconds story in my new MV document (https://docs.google.com/document/d/1yk7MVkXM3LEDV_sb9FMHCQmMJYYwGN6fwsz3YzbOEc0/edit#) and I'm not sure I understood the whole story. If I understand correctly, the reason why we don't want gc_grace_seconds=0 for MV is that in the current implementation, _hints_ get expired if they are older than gc_grace_seconds. But my question is: **view hints** are special, they are not ordinary hints for the base table. When we wrtite them, which expiration time do we use for them - the base's gc_grace_seconds or the view's? Do you know where the code that does this lives?

@nyh View hints are special in the sense that they are managed by the view hints manager object and not the standard hints manager:

https://github.com/scylladb/scylladb/blob/298cafff3537950f1b17b1fcf119e3d831bcd786/service/storage_proxy.hh#L266-L269

Those hints managers store hints in separate files and are send the hints completely independently of each other, but because they both are `db::hints::manager`s they share the same code.

Specifically, just before we attempt to send a mutation, we check whether it is older than gc_grace_seconds of the destination table and drop it if it is indeed older:

https://github.com/scylladb/scylladb/blob/298cafff3537950f1b17b1fcf119e3d831bcd786/db/hints/internal/hint_sender.cc#L294-L306

In this case, `m.s` will refer to view's schema.
